File: backend/scripts/resource_monitor.py
#!/usr/bin/env python3
"""
resource_monitor.py — PhysiOps Banking System
Uses psutil to check actual RAM and Disk usage on the self-hosted runner (your PC).

Exit codes:
  0 = Resources within limits  (GitHub Actions: step SUCCESS → build_success webhook)
  1 = Threshold exceeded        (GitHub Actions: step FAILURE → ram_warning webhook)

Thresholds:
  RAM  > 85%  → warning
  Disk > 90%  → warning
"""

import os
import sys
import json
import psutil

# ── Thresholds ────────────────────────────────────────────────────────────────
RAM_THRESHOLD_PERCENT  = 85.0
DISK_THRESHOLD_PERCENT = 90.0

# ── RAM ───────────────────────────────────────────────────────────────────────
ram = psutil.virtual_memory()
ram_used_gb  = round(ram.used  / (1024 ** 3), 2)
ram_total_gb = round(ram.total / (1024 ** 3), 2)
ram_percent  = round(ram.percent, 1)

# ── Disk ──────────────────────────────────────────────────────────────────────
# Use C:/ on Windows, / on Linux
disk_path = "C:\\" if sys.platform == "win32" else "/"
disk = psutil.disk_usage(disk_path)
disk_used_gb  = round(disk.used  / (1024 ** 3), 1)
disk_total_gb = round(disk.total / (1024 ** 3), 1)
disk_percent  = round(disk.percent, 1)

# ── Report ────────────────────────────────────────────────────────────────────
report = {
    "ram": {
        "usedGB":   ram_used_gb,
        "totalGB":  ram_total_gb,
        "percent":  ram_percent,
        "warning":  ram_percent > RAM_THRESHOLD_PERCENT,
    },
    "disk": {
        "usedGB":   disk_used_gb,
        "totalGB":  disk_total_gb,
        "percent":  disk_percent,
        "warning":  disk_percent > DISK_THRESHOLD_PERCENT,
    },
    "thresholds": {
        "ramPercent":  RAM_THRESHOLD_PERCENT,
        "diskPercent": DISK_THRESHOLD_PERCENT,
    },
}

print(json.dumps(report, indent=2))

# ── Decision ──────────────────────────────────────────────────────────────────
warnings = []
if report["ram"]["warning"]:
    warnings.append(f"RAM at {ram_percent}% ({ram_used_gb}/{ram_total_gb} GB)")
if report["disk"]["warning"]:
    warnings.append(f"Disk at {disk_percent}% ({disk_used_gb}/{disk_total_gb} GB)")

if warnings:
    if "GITHUB_OUTPUT" in os.environ:
        with open(os.environ["GITHUB_OUTPUT"], "a") as f:
            f.write(f"ram_percent={int(ram_percent)}\n")
    print(f"\n[WARNING] Resource threshold exceeded: {' | '.join(warnings)}", file=sys.stderr)
    print("[WARNING] Possible infinite loop or resource leak in this build!", file=sys.stderr)
    sys.exit(1)

if "GITHUB_OUTPUT" in os.environ:
    with open(os.environ["GITHUB_OUTPUT"], "a") as f:
        f.write(f"ram_percent={int(ram_percent)}\n")

print(f"\n[OK] RAM: {ram_percent}% | Disk: {disk_percent}% — within limits.")
sys.exit(0)

File: backend/scripts/resource-monitor.js
/**
 * resource-monitor.js
 * PhysiOps Banking System — RAM & Storage Monitor
 *
 * Checks current RAM/Disk usage on the runner machine.
 * Exit code 0 = all OK (used as "success" in GitHub Actions)
 * Exit code 1 = threshold exceeded (triggers ram_warning status)
 *
 * Thresholds:
 *   RAM  > 85%  → warning
 *   Disk > 90%  → warning
 */

const os = require('os');

async function checkDiskSpace(path) {
  try {
    // Try to use check-disk-space if installed
    const checkDiskSpace = require('check-disk-space').default;
    return await checkDiskSpace(path);
  } catch {
    // Fallback: return dummy data if package not available
    return { free: 50 * 1024 ** 3, size: 100 * 1024 ** 3 };
  }
}

async function main() {
  // ── RAM check ──────────────────────────────────────────────────────────────
  const totalRam   = os.totalmem();
  const freeRam    = os.freemem();
  const usedRam    = totalRam - freeRam;
  const ramPercent = ((usedRam / totalRam) * 100).toFixed(1);
  const ramUsedMB  = (usedRam / 1024 / 1024).toFixed(0);
  const ramTotalMB = (totalRam / 1024 / 1024).toFixed(0);

  // ── Disk check ─────────────────────────────────────────────────────────────
  const diskPath      = process.platform === 'win32' ? 'C:/' : '/';
  const disk          = await checkDiskSpace(diskPath);
  const diskUsedBytes = disk.size - disk.free;
  const diskPercent   = ((diskUsedBytes / disk.size) * 100).toFixed(1);
  const diskUsedGB    = (diskUsedBytes / 1024 / 1024 / 1024).toFixed(1);
  const diskTotalGB   = (disk.size / 1024 / 1024 / 1024).toFixed(1);

  // ── Output JSON ───────────────────────────────────────────────────────────
  const report = {
    ram: {
      usedMB: Number(ramUsedMB),
      totalMB: Number(ramTotalMB),
      usedPercent: Number(ramPercent),
      warning: Number(ramPercent) > 85,
    },
    disk: {
      usedGB: Number(diskUsedGB),
      totalGB: Number(diskTotalGB),
      usedPercent: Number(diskPercent),
      warning: Number(diskPercent) > 90,
    },
    threshold: {
      ramPercent: 85,
      diskPercent: 90,
    },
    triggeredAt: new Date().toISOString(),
  };

  console.log(JSON.stringify(report, null, 2));

  // ── Decision ───────────────────────────────────────────────────────────────
  if (report.ram.warning || report.disk.warning) {
    const msgs = [];
    if (report.ram.warning)  msgs.push(`RAM at ${ramPercent}% (${ramUsedMB}/${ramTotalMB} MB)`);
    if (report.disk.warning) msgs.push(`Disk at ${diskPercent}% (${diskUsedGB}/${diskTotalGB} GB)`);
    console.error(`\n⚠️  RESOURCE WARNING: ${msgs.join(' | ')}`);
    process.exit(1);   // → GitHub Actions marks this step as failed → triggers ram_warning webhook
  }

  console.log(`\n✅ Resources OK — RAM: ${ramPercent}% | Disk: ${diskPercent}%`);
  process.exit(0);
}

main().catch(err => {
  console.error('Resource monitor error:', err);
  process.exit(1);
});
